metaflow/runtime.py

```diff
@@ -798,56 +836,110 @@ def __init__(
+                self._is_resume_leader = False
+                resume_leader = None
+
+                if task_id_exists_already:
+                    # If the task id already exists, we need to check if current task is the resume leader in previous attempt.
+                    ds = self._flow_datastore.get_task_datastore(
+                        self.run_id, self.step, self.task_id
+                    )
+                    if not ds["_task_ok"]:
+                        raise MetaflowInternalError(
+                            "Externally cloned _parameters task did not succeed"
+                        )
+
+                    # Check if we should be the resume leader (maybe from previous attempt)
+                    if ds.has_metadata("_resume_leader", add_attempt=False):
+                        resume_leader = ds.load_metadata(
+                            ["_resume_leader"], add_attempt=False
+                        )["_resume_leader"]
+                        self._is_resume_leader = resume_leader == resume_identifier
+                else:
+                    # If the task id does not exist, current task is the resume leader.
+                    resume_leader = resume_identifier
+                    self._is_resume_leader = True
+
+                self.log(
+                    "Resume leader is %s." % resume_leader,
+                    system_msg=True,
+                )
+
+                if self._is_resume_leader:
+                    self.log(
+                        "Selected as the reentrant clone leader.",
+                        system_msg=True,
+                    )
                     # Clone in place without relying on run_queue.
                     self.new_attempt()
                     self._ds.clone(origin)
+                    # Set the resume leader be the task that calls the resume (first task to clone _parameters task).
+                    if resume_identifier:
+                        self._ds.save_metadata(
+                            {"_resume_leader": resume_identifier}, add_attempt=False
+                        )
+
                     self._ds.done()
                 else:
-                    # TODO: There is a bit of a duplication with the task.py
-                    # clone_only function here
-                    self.log(
-                        "Waiting for clone of _parameters step to occur...",
-                        system_msg=True,
-                    )
+                    # Wait for the resume leader to complete
                     while True:
-                        try:
-                            ds = self._flow_datastore.get_task_datastore(
-                                self.run_id, self.step, self.task_id
+                        ds = self._flow_datastore.get_task_datastore(
+                            self.run_id, self.step, self.task_id
+                        )
+                        if not ds["_task_ok"]:
```

Comment:
why error out so fast here? Shouldn't we just retry. Isn't this scenario possible:
- resume leader elected (task A)
- task B gets here before task A executes line 900

This would error out here but is not an error per say. We just need to try again.

Reply:
Thats a good point. I actually want to check if "_task_ok" is set and the value is False (meaning task failed) (maybe this doesn't do that?) let me check what is the default behavior of `ds["_task_ok"]`.

Also, thinking about it twice, I may not need to check this at all because if the leader fails, the other task will be stopped anyways. WDYT?

Reply:
removed